feat(common): display default template in NgForOf when empty

trotyl opened the pull request:
closes #14479

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #14479


## What is the new behavior?

`NgForOf` can now have a default template when no items available.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

Question:

+ What should be the property name? `default`, `else`, `empty`, `otherwise`, `fallback`...
+ Should **structural directives** guide being updated as well or leave that to separate PR?

mary-poppins commented:
You can preview 67d2704 at https://pr22275-67d2704.ngbuilds.io/.

trotyl commented:
API doc missing, will add soon.

vicb commented:
Great change, thanks !

We probably need to discuss the name (ie `default` vs `empty`) and if we should handle `| async` in this PR.

cl/186104809

trotyl commented on packages/common/src/directives/ng_for_of.ts:
```diff
@@ -148,7 +157,15 @@ export class NgForOf<T> implements DoCheck, OnChanges {
-      if (changes) this._applyChanges(changes);
+      if (changes) {
+        if (this._defaultBlockView) {
+          this._viewContainer.remove(this._viewContainer.indexOf(this._defaultBlockView));
```
If there's change, would be either:

+ 0 items => N items;
+ N items => 0 items;
+ N items => N items.

In first case, the removal is obvious right. In the other cases, there won't be an existing default block.
There there won't be a removal + create in one turn.

mary-poppins commented:
You can preview f225c04 at https://pr22275-f225c04.ngbuilds.io/.

mary-poppins commented:
You can preview ae0a1a5 at https://pr22275-ae0a1a5.ngbuilds.io/.

trotyl committed:
feat(common): display default template in NgForOf when empty

Add support for `default` template to be displayed when `*ngFor` has no
content.

closes #14479

mary-poppins commented:
You can preview 4b558cf at https://pr22275-4b558cf.ngbuilds.io/.

vicb commented:
Need discussions

lacolaco commented:
This is very cool and I think its name is important. 

`default` can mean it will be used if the source is  **null**  or **undefined** or **empty**. 
and it is not clear whether if it will turn back into _default view_ when the source has changed to empty 2nd and more times. ([] => [0 ,1] => []). Totally, I don't prefer the name `default` .

I propose `ifEmpty` for clarity and restrict it to be used when the source is NOT null and it is empty. (same to current implementation)

```html
<ng-container *ngFor="let item of items$ | async; ifEmpty: noItems">
</ng-container>

<ng-template #noItems>
  There are no Items.
 </ng-template>
```

jessicajaniuk commented on packages/common/src/directives/ng_for_of.ts:
```diff
@@ -113,8 +118,17 @@ export class NgForOf<T> implements DoCheck, OnChanges {
 
+  @Input()
+  set ngForDefault(value: TemplateRef<void>) {
+    if (value) {
```
is it possible to single line this? Since _defaultBlockTemplate is `TemplateRef<void>|null`, wouldn't `this._defaultBlockTemplate = value ?? null;` work?

mhevery replied:
fixed

jessicajaniuk commented on packages/common/test/directives/ng_for_spec.ts:
```diff
@@ -359,6 +359,26 @@ let thisArg: any;
+        getComponent().items = [];
+        detectChangesAndExpectText('No items');
+
+        getComponent().items = [];
```
This duplicates Lines 375 and 376. Is there a reason for this or can the duplicates be removed?

petebacondarwin replied:
I agree, these look superfluous.

jessicajaniuk reviewed (approved):
LGTM; Just a couple of minor comments.

mhevery commented:
[presubmit](http://test/OCL:345559540:BASE:345559560:1607041416264:3a9d453b)

lacolaco commented on goldens/public-api/common/common.d.ts:
```diff
@@ -215,6 +215,7 @@ export declare class NgComponentOutlet implements OnChanges, OnDestroy {
 }
 
 export declare class NgForOf<T, U extends NgIterable<T> = NgIterable<T>> implements DoCheck {
+    set ngForDefault(value: TemplateRef<void>);
```
wrong name?

mhevery replied:
thanks, fixed

mary-poppins commented:
You can preview 2b2f16c at https://pr22275-2b2f16c.ngbuilds.io/.

mary-poppins commented:
You can preview 025cdba at https://pr22275-025cdba.ngbuilds.io/.

mary-poppins commented:
You can preview 0027e09 at https://pr22275-0027e09.ngbuilds.io/.

petebacondarwin commented on packages/common/src/directives/ng_for_of.ts:
```diff
@@ -180,6 +180,23 @@ export class NgForOf<T, U extends NgIterable<T> = NgIterable<T>> implements DoCh
+   */
+  @Input()
+  set ngForWhenEmpty(value: TemplateRef<void>) {
+    this._whenEmptyBlockTemplate = value ?? null;
```
In `ngIf` it is assumed that equivalent inputs (e.g. `ngfElse` and `ngIfThen`) can change and so it triggers an update to the view whenever their setter is called. Should we do that here too?

petebacondarwin replied:
I also note that `ngIf` does a runtime check that the value is actually a template (by calling a helper e.g. `assertTemplate('ngIfElse', templateRef);`). But I assume that now we have much tighter type checking via Ivy this is no longer needed.

petebacondarwin reviewed (approved):
I am a bit concerned about the extra payload cost but will go with the consensus opinion.
One other query, which is not blocking.
And I agree that @jessicajaniuk's comment should be addressed.

Otherwise LGTM!
Reviewed-for: public-api

jelbourn commented:
Hey all, despite some early approvals from team members, after further discussion we've decided not to add this feature. The main rationale here is that this can already be accomplished by using `ngIf` after the `ngFor` without any additional APIs. The proposed API would add ~500 bytes to the framework for all users. While this is small, many small features like this add up to kilobytes over time. Expanding the API surface also increases the amount of content for the docs, meaning new users have more in front of them when trying to learn concepts. 

I know that there are a number of people who would find this feature useful, but ultimately we think it's to the benefit of the larger audience not add this one.
